server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -3095,6 +3112,13 @@ private EngineConfig newEngineConfig(LongSupplier globalCheckpointSupplier) {
             }
         };
 
+        final List<ReferenceManager.RefreshListener> internalRefreshListener;
+        if (indexSettings.isSegRepEnabled() && shardRouting.primary()) {
+            internalRefreshListener = Arrays.asList(new RefreshMetricUpdater(refreshMetric), checkpointRefreshListener);
+        } else {
+            internalRefreshListener = Collections.singletonList(new RefreshMetricUpdater(refreshMetric));
```

Comment:
This isn't used anywhere? It should be wired up in `newEngineConfig` below.

Should this be an internal or external listener?

Reply:
Lets push this down to the creation of `IndexShard` so that we can minimize the branching across the board and create the right flavour of `IndexShard` for segrep

Reply:
@kartg this we how we have it now, The above highlighted one is outdated.

```
final List<ReferenceManager.RefreshListener> internalRefreshListener;
        if (this.checkpointRefreshListener != null) {
            internalRefreshListener = Arrays.asList(new RefreshMetricUpdater(refreshMetric), checkpointRefreshListener);
        } else {
            internalRefreshListener = Collections.singletonList(new RefreshMetricUpdater(refreshMetric));
        }
```

Also if we try to write it as your suggested way, then internalRefreshListeners cannot be final, we have been using it as final even before this PR. Not sure if changing internalRefreshListener to a non-final variable is a good idea.

Reply:
Given that this block declares and defines the `internalRefreshListener` variable, the `final` modifier doesn't add value IMO. What is more valuable to maintain is the _unmodifiable list_, which is lost if `checkpointRefreshListener` is non-null. 

If you'd really like to retain the `final` modifier, I'd suggest doing something like this:

```
List<ReferenceManager.RefreshListener> refreshListenerList = Arrays.asList(new RefreshMetricUpdater(refreshMetric));
if (this. checkpointRefreshListener != null) {
    refreshListenerList.add(checkpointRefreshListener);
}
final List<ReferenceManager.RefreshListener> internalRefreshListener = Collections.unmodifiableList(refreshListenerList);
```

Reply:
@kartg I tried to incorporate your code block, but it throws me UnsupportedOperation Exception when we are adding, `"refreshListenerList.add(checkpointRefreshListener);"`. As we are using `Arrays.asList(new RefreshMetricUpdater(refreshMetric));` we cannot use .add method. Instead we have to again convert this into actual List.  More about this exception can be found [here](https://www.baeldung.com/java-list-unsupported-operation-exception). As this is a nitpick I am just using the way how it was before to avoid complexity